src/client/ua_client_subscriptions.c

```diff
@@ -21,6 +21,47 @@
 /* Subscriptions */
 /*****************/
 
+static UA_StatusCode __Subscriptions_create_prepare(CustomCallback *cc, const UA_CreateSubscriptionRequest *request,
+                                                    void *subscriptionContext,
+                                                    UA_Client_StatusChangeNotificationCallback statusChangeCallback,
+                                                    UA_Client_DeleteSubscriptionCallback deleteCallback) {
+    UA_Client_Subscription *sub =
+        (UA_Client_Subscription *)(cc->clientData = UA_malloc(sizeof(UA_Client_Subscription)));
+    cc->clientDataDeleter = UA_free;
+    if (!sub)
+        return UA_STATUSCODE_BADOUTOFMEMORY;
+    sub->context = subscriptionContext;
+    sub->statusChangeCallback = statusChangeCallback;
+    sub->deleteCallback = deleteCallback;
+    return UA_STATUSCODE_GOOD;
+}
+
+static void __Subscriptions_create_handler(UA_Client *client, void *data, UA_UInt32 requestId, void *r) {
+    UA_CreateSubscriptionResponse *response = (UA_CreateSubscriptionResponse *)r;
+    CustomCallback *cc = (CustomCallback *)data;
+    if (response->responseHeader.serviceResult != UA_STATUSCODE_GOOD) {
+        goto cleanup;
+    }
+
+    UA_Client_Subscription *newSub = (UA_Client_Subscription *)cc->clientData;
+    cc->clientData = NULL;
+
+    /* Prepare the internal representation */
+    newSub->subscriptionId = response->subscriptionId;
+    newSub->sequenceNumber = 0;
+    newSub->lastActivity = UA_DateTime_nowMonotonic();
+    newSub->publishingInterval = response->revisedPublishingInterval;
+    newSub->maxKeepAliveCount = response->revisedMaxKeepAliveCount;
+
+    LIST_INIT(&newSub->monitoredItems);
+    LIST_INSERT_HEAD(&client->subscriptions, newSub, listEntry);
+
+cleanup:
+    if (cc->userCallback)
+        cc->userCallback(client, cc->userData, requestId, response);
+    CustomCallback_remove(cc, false);
```

Comment:
The handler should not have to remove the callback.
This can be done "outside" in the generic handling of async callbacks.

Then, we can also stack-allocate the CustomCallback when call the handler in the synchronous case.

Reply:
I'm not happy with my approach to delete it directly in the handler either (I previously tried to stack allocate on synchronous functions too). Didnt get it to work properly.

I could add the routine to automatically free a stored customcallback using the requestId in processAsyncMessage (ua_client.c:241). However this would require for every async callback to check (find, delete, remove from list) if there is a customcallback stored, even if there is none saved.

What do you think?

Reply:
see first comment (CustomCallbackDataDeleter)